### DISCUSSION TITLE
Pass properties to custom tools

### DISCUSSION THREAD

**Discussion by formula23** [FlowiseAI/Flowise#1168] [Q&A]
We have created a few custom tools in our flow. We would like to pass custom variables in our API request. But we can figure out how to send with the overrideConfig? Is this possible?

**Comment by HenryHengZJ** [comment]
not possible yet. We are working on that to allow user pass env variable so you can use it within custom tool as `env.{variable}`

**Comment by formula23** [comment]
Thanks Henry, we were trying to pass custom properties to the custom tool from the API and pass in overrideConfig. Would be great to create custom properties on a custom tool that can be overwritten with the API request.

**Reply by Jaredude** [reply]
@HenryHengZJ env.{variable} will be very helpful with being able to allow authenticated API calls in a custom tool! It would also be great if we could pass in some sort of custom properties via the Chatbot's .init and initFull call. Something along the lines of 
```
Chatbot.init({
  chatflowid,
  apiHost,
  customProperties: {
    [key]: value
  }
})

```
where we could then digest it within the custom tool in a similar way to what you're talking about with an env object. A use case I could see this being very helpful with would be to allow the chatbot to act as a proxy on behalf of an authenticated user.

For example, we could pass an auth token that an authenticated user has in localStorage into the chatbot init. The chatbot could then use that auth token to make an authenticated request on behalf of the user. A real world usage would be that that I'm logged into a website and the website has a bot that handles updates to my profile so that I don't have to navigate around to different pages update my profile, add new products, etc. I could tell the bot something like: Please add a new set of 21" Arachnid rims to my profile at the price of $2,500USD. The bot could then make an authenticated API request on my behalf to do so.

I'd be happy to contribute code, testing, brainstorming on adding a feature like this if there's interest in making it happen. I'm just getting into flowise and absolutely loving what is capable of already!

**Reply by HenryHengZJ** [reply]
@Jaredude @formula23 We now allow passing `vars` in your API that can be used in custom tool: https://docs.flowiseai.com/integrations/tools/custom-tool#pass-variables-to-function

**Reply by Jaredude** [reply]
> @Jaredude @formula23 We now allow passing `vars` in your API that can be used in custom tool: https://docs.flowiseai.com/integrations/tools/custom-tool#pass-variables-to-function
> 
> 
> 
> 

Works great! Such a wonderful feature add! Thanks!

**Comment by dkindlund** [comment]
@HenryHengZJ , while you're at it, could you also allow code in Custom Tools to access the default `{{question}}` and `{{chat_history}}` values?  I'm having a really hard time getting that data consistently passed into my Custom Tool.  Any ideas?

**Reply by dkindlund** [reply]
Oh, my mistake, it looks like `{{question}}` is accessible from `$flow.input`.  But I don't see an equivalent `$flow.chat_history` for `{{chat_history}}`

**Reply by HenryHengZJ** [reply]
@dkindlund we can look at how do do that, for now you can actually get the chat history by using the `$flow.sessionId` or `$flow.chatId` with combination of `$flow.chatflowId` and make a call to Message API - https://docs.flowiseai.com/using-flowise/api#message-api

**Reply by dkindlund** [reply]
Got it!  Thank you very much for the additional pointers here, @HenryHengZJ -- I totally forgot about the other APIs until you mentioned them.
